bundles/org.openhab.binding.teslascope/src/main/java/org/openhab/binding/teslascope/internal/api/DetailedInformation.java

```diff
@@ -0,0 +1,247 @@
+ * @author Paul Smedley - Initial Contribution
+ *
+ */
+@NonNullByDefault
+public class DetailedInformation {
+    public String vin = "";
+    public String vehicleName = "";
+    public String vehicleState = "";
+    public double odometer;
+
+    // charge_state
+    public int batteryLevel;
+    public int usableBatteryLevel;
+    public float batteryRange;
+    public float estBatteryRange;
+    public boolean chargeEnableRequest = false;
+    public float chargeEnergyAdded;
+    public int chargeLimitSoc;
+    public boolean chargePortDoorOpen = false;
+    public float chargeRate;
+    public int chargerPower;
+    public int chargerVoltage;
+    public String chargingState = "";
+    public float timeToFullCharge;
+    public boolean scheduledChargingPending = false;
+    public String scheduledChargingStartTime = " ";
+
+    // climate_state
+    public boolean isAutoConditioningOn;
+    public boolean isClimateOn;
+    public boolean isFrontDefrosterOn;
+    public boolean isPreconditioning;
+    public boolean isRearDefrosterOn;
+    public int seatHeaterLeft;
+    public int seatHeaterRearCenter;
+    public int seatHeaterRearLeft;
+    public int seatHeaterRearRight;
+    public int seatHeaterRight;
+    public boolean sideMirrorHeaters;
+    public boolean smartPreconditioning;
+    public boolean steeringWheelHeater;
+    public boolean wiperBladeHeater;
+    public float driverTempSetting;
+    public float insideTemp;
+    public float outsideTemp;
+    public float passengerTempSetting;
+    public float fanStatus;
+    public int leftTempDirection;
+    public float maxAvailTemp;
+    public float minAvailTemp;
+    public int rightTempDirection;
+
+    // drive state
+    public int heading;
+    public float latitude;
+    public float longitude;
+    public String shiftState = "";
+    public float power;
+    public float speed = 0;
+
+    // vehicle_state
+    public boolean locked;
+    public boolean sentryMode;
+    public boolean valetMode;
+    public String softwareUpdateStatus = "";
+    public String softwareUpdateVersion = "";
+    public boolean fdWindow;
+    public boolean fpWindow;
+    public boolean rdWindow;
+    public boolean rpWindow;
+    public String sunRoofState = "";
+    public int sunRoofPercentOpen;
+    public boolean homelinkNearby;
+    public double tpmsPressureFL;
+    public double tpmsPressureFR;
+    public double tpmsPressureRL;
+    public double tpmsPressureRR;
+    public boolean tpmsSoftWarningFL;
+    public boolean tpmsSoftWarningFR;
+    public boolean tpmsSoftWarningRL;
+    public boolean tpmsSoftWarningRR;
+    public boolean df;
+    public boolean dr;
+    public boolean pf;
+    public boolean pr;
+    public boolean ft;
+    public boolean rt;
+
+    private DetailedInformation() {
+    }
+
+    public static DetailedInformation parse(String response) {
+        /* parse json string */
+        JsonObject jsonObject = JsonParser.parseString(response).getAsJsonObject();
```

Comment:
Without looking into all details, why not prevent all this plumbing and parsing just by using something like:
```
return gson.fromJson(response, DetailedInformation.class);
```

Reply:
I didn't know about this :) That would have saved me a bunch of pain! I'll make this update ASAP

Reply:
OK I've implemented this, let me do some testing to see what broke before I ask for a re-review :)

Reply:
Yep, I broke it  - solution isn't as simple as I first thought.  I get a single response string like:
`{"id":19919,"user_id":22xxx,"tesla_account_id":xxxx,"vehicle_id":xxxxxxxxxxxxxxx,"special_vehicle_id":xxxxxxxxxxxxxxx,"public_id":"xxxx","vin":"LRWYHCFS8PCxxxxx","name":"Tessy","state":"asleep","api_version":73,"primary":1,"logging":1,"version_logging":1,"idle_override":15,"recent_override":10,"free_supercharging":0,"fsd_enabled":0,"eap_enabled":1,"ap_enabled":1,"calculate_cost":0,"allow_unlocking":1,"shared_override":1,"billing_day":1,"private":1,"controls":0,"dashboard_type":0,"country":"au","restful":1,"sleep_start_at":null,"sleep_end_at":null,"ready":1,"co2_saved":4,"trees_planted":93,"safety_logging":1,"specs":"[{\"code\":\"$INPB0\",\"colorCode\":\"\",\"displayName\":\"All Black Premium Interior\",\"endDate\":\"\",\"isActive\":true},{\"code\":\"$PPSW\",\"colorCode\":\"PPSW\",\"displayName\":\"Pearl White Multi-Coat\",\"endDate\":\"\",\"isActive\":true},{\"code\":\"$WY20P\",\"colorCode\":\"\",\"displayName\":\"20\\u2019\\u2019 Induction Wheels\",\"endDate\":\"\",\"isActive\":true},{\"code\":\"$STY5S\",\"colorCode\":\"\",\"displayName\":\"Five Seat Interior\",\"endDate\":\"\",\"isActive\":true},{\"code\":\"$MTY13\",\"colorCode\":\"\",\"endDate\":\"\",\"isActive\":true},{\"code\":\"$APBS\",\"colorCode\":\"\",\"displayName\":\"Autopilot\",\"endDate\":\"\",\"isActive\":true},{\"code\":\"$APPB\",\"colorCode\":\"\",\"displayName\":\"Enhanced Autopilot\",\"endDate\":\"\",\"isActive\":true},{\"code\":\"$SC04\",\"colorCode\":\"\",\"displayName\":\"Supercharger Network Access + Pay-as-you-go\",\"endDate\":\"\",\"isActive\":true}]","created_at":"2023-08-02T10:18:30.000000Z","updated_at":"2024-07-15T07:11:14.000000Z","recently_updated_at":"2024-07-15T07:27:14.000000Z","image_url":null,"recently_driven_at":"2024-07-15T07:11:14.000000Z","recently_charged_at":"2024-07-13T04:52:17.000000Z","odometer":"4928.774318","fleet_enabled":1,"forced_sentry":0,"access_type":"OWNER","virtual_key":1,"recently_updated":false,"vehicle_config":{"id":19346,"vehicle_id":19919,"can_accept_navigation_requests":1,"can_actuate_trunks":1,"car_special_type":"base","car_type":"modely","charge_port_type":"CCS","eu_vehicle":1,"mcu":2,"exterior_color":"PearlWhite","exterior_trim":"Black","has_air_suspension":0,"has_ludicrous_mode":0,"key_version":2,"motorized_charge_port":1,"perf_config":"","plg":"1","rear_seat_heaters":1,"roof_color":"RoofColorGlass","spoiler_type":"None","trim_badging":"50","wheel_type":"Induction20Black","created_at":"2023-08-02T10:24:08.000000Z","updated_at":"2024-06-12T07:50:12.000000Z","driver_assist":"TeslaAP3","interior_trim_type":"Black2"},"vehicle_state":{"id":19356,"vehicle_id":19919,"car_version":"2024.20.9 1a02147be9af","hw_version":"3.0","is_user_present":0,"locked":1,"odometer":"4928.774318","sentry_mode":0,"speed_limit_mode":0,"valet_mode":0,"current_limit_mph":85,"max_limit_mph":120,"min_limit_mph":50,"software_update_status":"","software_update_version":" ","software_update_expected_duration":2700,"software_update_download_perc":0,"software_update_install_perc":1,"fd_window":0,"fp_window":0,"rd_window":0,"rp_window":0,"tpms_pressure_fl":"2.725","tpms_pressure_fr":"2.75","tpms_pressure_rl":"2.75","tpms_pressure_rr":"2.775","sun_roof_state":null,"sun_roof_percent_open":null,"homelink_device_count":3,"homelink_nearby":1,"created_at":"2023-08-02T10:24:08.000000Z","updated_at":"2024-07-15T07:27:14.000000Z","dashcam_state":null,"tpms_rcp_front_value":"2.900","tpms_rcp_rear_value":"2.900","tpms_hard_warning_fl":0,"tpms_hard_warning_fr":0,"tpms_hard_warning_rl":0,"tpms_hard_warning_rr":0,"tpms_soft_warning_fl":0,"tpms_soft_warning_fr":0,"tpms_soft_warning_rl":0,"tpms_soft_warning_rr":0,"df":0,"dr":0,"pf":0,"pr":0,"ft":0,"rt":0},"climate_state":{"id":19349,"vehicle_id":19919,"driver_temp_setting":"23.00","passenger_temp_setting":"23.00","fan_status":"0.00","inside_temp":"18.50","outside_temp":"11.50","max_avail_temp":"28.00","min_avail_temp":"15.00","is_auto_conditioning_on":0,"is_climate_on":0,"is_front_defroster_on":0,"is_rear_defroster_on":0,"is_preconditioning":0,"smart_preconditioning":0,"left_temp_direction":0,"right_temp_direction":0,"steering_wheel_heater":0,"seat_heater_left":0,"seat_heater_right":0,"seat_heater_rear_center":0,"seat_heater_rear_left":0,"seat_heater_rear_right":0,"side_mirror_heaters":0,"wiper_blade_heater":0,"created_at":"2023-08-02T10:24:08.000000Z","updated_at":"2024-07-13T05:48:23.000000Z","climate_keeper_mode":"off"},"charge_state":{"id":19354,"vehicle_id":19919,"battery_level":84,"usable_battery_level":84,"battery_range":"221.66","est_battery_range":"213.17","charge_enable_request":1,"charge_energy_added":"22.10","charge_limit_soc":100,"charge_limit_soc_max":100,"charge_limit_soc_min":50,"charge_limit_soc_std":80,"charge_port_cold_weather_mode":0,"charge_port_door_open":0,"charge_port_latch":"Engaged","charge_rate":"0.00","charger_power":0,"charger_voltage":2,"charging_state":"Disconnected","minutes_to_full_charge":0,"time_to_full_charge":"0.00","conn_charge_cable":"<invalid>","fast_charger_brand":"<invalid>","fast_charger_type":"<invalid>","fast_charger_present":"","scheduled_charging_pending":0,"scheduled_charging_start_time":null,"created_at":"2023-08-02T10:24:08.000000Z","updated_at":"2024-07-13T05:48:23.000000Z"},"drive_state":{"id":19351,"vehicle_id":19919,"heading":358,"latitude":"-34.90879200","longitude":"138.49962100","native_location_supported":1,"native_latitude":"-xxxxxxxx","native_longitude":"xxxxxxx","native_type":"wgs","power":0,"shift_state":null,"speed":null,"created_at":"2023-08-02T10:24:08.000000Z","updated_at":"2024-07-13T05:48:23.000000Z"},"locations":[{"id":7099,"vehicle_id":19919,"category":"default","name":"Dad's","description":null,"latitude":"-xxxxx","longitude":"xxxx","radius":"0.50","area":null,"geocode_id":0,"is_present":false,"notify_on_enter":true,"notify_on_leave":true,"created_at":"2023-08-07T08:37:45.000000Z","updated_at":"2023-08-07T08:37:45.000000Z","triggered_at":null},{"id":7149,"vehicle_id":19919,"category":"default","name":"Dinga","description":null,"latitude":"-xxxxx","longitude":"xxxxx","radius":"0.50","area":null,"geocode_id":0,"is_present":false,"notify_on_enter":true,"notify_on_leave":true,"created_at":"2023-08-18T09:35:16.000000Z","updated_at":"2023-08-18T09:35:16.000000Z","triggered_at":null},{"id":7150,"vehicle_id":19919,"category":"default","name":"Grammy's","description":null,"latitude":"-xxxxx","longitude":"xxxxx","radius":"0.50","area":null,"geocode_id":0,"is_present":false,"notify_on_enter":true,"notify_on_leave":true,"created_at":"2023-08-18T09:35:47.000000Z","updated_at":"2023-08-18T09:35:47.000000Z","triggered_at":null},{"id":7097,"vehicle_id":19919,"category":"default","name":"Home","description":null,"latitude":"-xxxxx","longitude":"xxxxx","radius":"0.50","area":null,"geocode_id":0,"is_present":false,"notify_on_enter":true,"notify_on_leave":true,"created_at":"2023-08-07T08:37:04.000000Z","updated_at":"2023-08-07T08:37:04.000000Z","triggered_at":null},{"id":7098,"vehicle_id":19919,"category":"default","name":"Work","description":null,"latitude":"-xxxxx","longitude":"xxxxx","radius":"0.50","area":null,"geocode_id":0,"is_present":false,"notify_on_enter":true,"notify_on_leave":true,"created_at":"2023-08-07T08:37:26.000000Z","updated_at":"2023-08-07T08:37:26.000000Z","triggered_at":null}],"safety_score":null,"data_request":{"id":77214,"vehicle_id":19919,"date":"2024-07-15T00:00:00.000000Z","count":33,"command_count":0,"created_at":"2024-07-15T06:40:25.000000Z","updated_at":"2024-07-15T06:40:25.000000Z"}}`

I probably need to break detailedinformation.java into individual classes for the 'groups'  of data that are parsed.

Reply:
You have to mimic the json structure in your dto classes, yes.
There are multiple tools that can help you with this. These dto classes can be generated from json.

Reply:
OK this is now fixed.